velox/connectors/hive/SplitReader.cpp

```diff
@@ -276,20 +276,44 @@ void SplitReader::createRowReader(
     std::shared_ptr<common::MetadataFilter> metadataFilter) {
   auto& fileType = baseReader_->rowType();
   auto columnTypes = adaptColumns(fileType, baseReaderOpts_.getFileSchema());
+  auto columnNames = fileType->names();
 
+  setRowIndexColumnInfoIfNeed(fileType, columnNames, columnTypes);
   configureRowReaderOptions(
       baseRowReaderOpts_,
       hiveTableHandle_->tableParameters(),
       scanSpec_,
       metadataFilter,
-      ROW(std::vector<std::string>(fileType->names()), std::move(columnTypes)),
+      ROW(std::vector<std::string>(columnNames), std::move(columnTypes)),
       hiveSplit_);
   // NOTE: we firstly reset the finished 'baseRowReader_' of previous split
   // before setting up for the next one to avoid doubling the peak memory usage.
   baseRowReader_.reset();
   baseRowReader_ = baseReader_->createRowReader(baseRowReaderOpts_);
 }
 
+void SplitReader::setRowIndexColumnInfoIfNeed(
+    const RowTypePtr& fileType,
+    std::vector<std::string>& columnNames,
+    std::vector<facebook::velox::TypePtr>& columnTypes) {
+  auto rowIndexMetaColIdx =
+      readerOutputType_->getChildIdxIfExists(kSparkReservedTmpMetaRowIndex);
+  if (rowIndexMetaColIdx.has_value() &&
+      !fileType->containsChild(kSparkReservedTmpMetaRowIndex) &&
+      hiveSplit_->partitionKeys.find(kSparkReservedTmpMetaRowIndex) ==
+          hiveSplit_->partitionKeys.end()) {
+    dwio::common::RowNumberColumnInfo rowNumberColumnInfo;
+    rowNumberColumnInfo.insertPosition = rowIndexMetaColIdx.value();
+    rowNumberColumnInfo.name = kSparkReservedTmpMetaRowIndex;
+    baseRowReaderOpts_.setRowNumberColumnInfo(rowNumberColumnInfo);
+    auto rowIndexColumnScanSpec =
+        scanSpec_->childByName(kSparkReservedTmpMetaRowIndex);
+    if (rowIndexColumnScanSpec) {
+      scanSpec_->removeChild(rowIndexColumnScanSpec);
```

Comment:
You need to avoid creating `ScanSpec` node for this column in `HiveDataSource`.  `ScanSpec` in general should not be mutated at split level.

Reply:
got, i add a new column handle type `HiveColumnHandle::ColumnType::kRowIndex` to let `HiveDataSource` can skip using this column built scanSpec in `makeScanSpec` function.

upstream engine is responsobile to set all columnHandles including row index column handle when init `TableScanNode`